### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] False positive when using generic parameter with default

### ISSUE THREAD

**Issue by srmagura** [typescript-eslint/typescript-eslint#5487]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.4&sourceType=module&code=C4TwDgpgBAggxsAlgewHYB4AqA+KBeKAbyjgENgIBzZAJxAC4pMoBfAKA4DMBXVBFVFAC2IAGK84WfFFTchAIwg1sAClL80AZ0bwkaLNgDaAXQCURNlBJbgUUjRqkQMB0+mH1e1JuMcr9x2dXEAA6TVpgFTUAGih5czxcQksra29bUgBJCiFpUkMABl9UtM1beWyIXIJ5Qt8UqwB6RqYAC0RNKAAbRFRoMBpkABNuOAhtBqhmyeaoACYAZgA2OcZMVuguiE5gAFpW0lQhqE1EIehkTjtBe0RgVqEIJDgoZEhHL2FuMriLq9BIFAAOSHEBA6IzFpA2QKJTg4HyRCUXrAIGvGjXKAQGFQAEQEKQyFQGhPbg0G6VIQhMgUah0KC7OKU6nkKi0EB+Eok4BkwRqZk0tn00idUHmRkqCo5Fm09l2UWoECmFIsZXsIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
type Action<T> = { category: T }


function myFunc<T = number>(actions: Action<T>[]) {
  const arrayArray = [actions]

  arrayArray.sort((a, b) => {
    const aItem = a[0]
    const bItem = b[0]

    // This line produces:
    //
    // 2362: The left-hand side of an arithmetic operation must be of type 'any',
    // 'number', 'bigint' or an enum type.
    //
    // return aItem.category - bItem.category

     return (aItem.category as any) - (bItem.category as any)
  })
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

There should be no linting errors, since the `as any` type assertions are necessary.

### Actual Result

The `as any` type assertions are reported as unnecessary.

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.33.1` |
| `@typescript-eslint/parser` | `5.33.1` |
| `TypeScript` | `4.7.4` |
| `ESLint` | `8.15.0` |
| `node` | `web` |

**Comment by bradzacher** [comment]
Your `T` has a default value of `number`, but has no constraint on it.
This means that the type `T` is an unconstrained type, which functionally behaves like `any`.

Given how your code is written (you assume it's safe to use `T` in a numeric operation) - you really want `T` to be a subtype of `number`.

So your code should be `T extends number = number`.
[If you do this then you don't need the assertion at all.](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggxsAlgewHYB4AqA+KBeKAbyjgENgIBzZAJxAC4pMoBfAKA4DMBXVBFVFAC2IAGK84WKBAAeFVABMAzlFTchAIwg1sAClL80SxvCRos2ANoBdAJRE2UEkeBRSNGqRAwPX-FEsDM1Qlaw4nd09vXxAAOiVaYF19ABooDXs8XEJHJ2cQ11IASQohf1JLAAYwvKgAejraqAA9AH5cpzgXdJKIMoINKpq8hqa28JHGzAALRBUAG0RUaDAaZAVuOAhjDvq63dGAJgBmADZDxhnoeYhOYABaadJFKCVEBWhkTjdBd0RgaZCCBIOBQZCQTzBYTcJSuLRg76gSBQADkzxAKJSB0aKLUmm0mNRGkQlCWwBRYJoP2keKgSIgsWx2KgNGB3Bov16QliZAo1DoUHuPVKPPIVFoIAmeVZwHZgn0XNFfIl9iFug0it54rotl2eyaBpa7ScLF1LCAA)
